apps/omg_watcher_info/lib/omg_watcher_info/api/transaction.ex

```diff
@@ -126,6 +188,78 @@ defmodule OMG.WatcherInfo.API.Transaction do
 
     outputs
     |> Stream.concat(Stream.repeatedly(empty_gen))
-    |> (&Enum.zip([:output0, :output1, :output2, :output3], &1)).()
+    |> (fn output -> Enum.zip([:output0, :output1, :output2, :output3], output) end).()
+  end
+
+  defp create_transaction(utxos_per_token, %{
+         owner: owner,
+         payments: payments,
+         metadata: metadata,
+         fee: fee
+       }) do
+    rests =
+      utxos_per_token
+      |> Stream.map(fn {token, utxos} ->
+        outputs =
+          [fee | payments]
+          |> Stream.filter(fn %{currency: currency} -> currency == token end)
+          |> Stream.map(fn %{amount: amount} -> amount end)
+          |> Enum.sum()
+
+        inputs = utxos |> Stream.map(fn %{amount: amount} -> amount end) |> Enum.sum()
+        %{amount: inputs - outputs, owner: owner, currency: token}
+      end)
+      |> Enum.filter(fn %{amount: amount} -> amount > 0 end)
+
+    outputs = payments ++ rests
+
+    inputs =
+      utxos_per_token
+      |> Enum.map(fn {_, utxos} -> utxos end)
+      |> List.flatten()
+
+    cond do
+      Enum.count(outputs) > Transaction.Payment.max_outputs() ->
+        {:error, :too_many_outputs}
+
+      Enum.empty?(inputs) ->
+        {:error, :empty_transaction}
```

Comment:
this whole thing really looks like it's a subset of some larger functionality it would make sense to extract into a separate module perhaps